src/workspaces/common/state/useWorkspaceStatePolling.ts

```diff
@@ -0,0 +1,121 @@
+import { LoadedState } from '@terra-ui-packages/core-utils';
+import _ from 'lodash/fp';
+import { useEffect, useRef, useState } from 'react';
+import { Ajax } from 'src/libs/ajax';
+import { workspacesStore, workspaceStore } from 'src/libs/state';
+import { pollWithCancellation } from 'src/libs/utils';
+import { BaseWorkspaceInfo, WorkspaceState, WorkspaceWrapper as Workspace } from 'src/workspaces/utils';
+
+// type WorkspaceUpdate = Partial<Workspace['workspace']>
+
+interface WorkspaceUpdate extends Partial<BaseWorkspaceInfo> {
+  workspaceId: string;
+  state: WorkspaceState;
+  errorMessage?: string;
+}
+
+const updateWorkspacesList = (workspaces: Workspace[], update: WorkspaceUpdate): Workspace[] =>
+  workspaces.map((ws) => updateWorkspaceIfMatching(ws, update));
+
+const updateWorkspaceIfMatching = <T extends Workspace>(workspace: T, update: WorkspaceUpdate): T => {
+  if (workspace.workspace?.workspaceId === update.workspaceId) {
+    const updatedWs = _.cloneDeep(workspace);
+    updatedWs.workspace.state = update.state;
```

Comment:
Nit: you might be able to use _.merge (given that you are already using _.cloneDeep).

Reply:
Though perhaps this is just clearer… nevermind.